File: problems/0008_string-to-integer-atoi/built-in/Solution.cpp
using namespace std;

class Solution {
public:
    int myAtoi(string str) {
        try {
            const long long result = stoll(str);
            if (result < INT_MIN)
                return INT_MIN;
            else if (result > INT_MAX)
                return INT_MAX;
            else
                return (int) result;
        } catch (invalid_argument& ignored) {
        } catch (out_of_range& ignored) {
            const size_t sLen = str.length();
            for (size_t i = 0; i < sLen; ++i) {
                const char c = str[i];
                if (c == '-')
                    return INT_MIN;
                else if (c == '+' || '0' <= c && c <= '9')
                    return INT_MAX;
            }
        }
        return 0;
    }
};

File: problems/0008_string-to-integer-atoi/built-in/solution.rb
#!/usr/bin/ruby

# @param {String} str
# @return {Integer}
def my_atoi(str)
    if match = str.match(/^\s*(([\-\+]?)(0*)(\d+)).*$/)
        begin
            _, s, z, n = match.captures
            result = Integer(s + n)
            if result < INT_MIN
                return INT_MIN
            elsif result > INT_MAX
                return INT_MAX
            else
                return result
            end
        rescue ArgumentError => ignored
        end
    end
    return 0;
end

INT_MAX = 2**31 - 1
INT_MIN = -2**31
